src/Compilers/CSharp/Test/Semantic/Semantics/NullCheckedParameterTests.cs

```diff
@@ -170,5 +177,318 @@ class C
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+            var m = comp.GlobalNamespace.GetTypeMember("C").GetMember<SourceMethodSymbol>("M");
+            Debug.Assert(((SourceParameterSymbol)m.Parameters[0]).IsNullChecked);
+        }
+
+        [Fact]
+        public void NullCheckedMethodValidationWithOptionalStringLiteralParameter()
+        {
+            var source = @"
+class C
+{
+    void M(string name! = ""rose"") { }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+            var m = comp.GlobalNamespace.GetTypeMember("C").GetMember<SourceMethodSymbol>("M");
+            Debug.Assert(((SourceParameterSymbol)m.Parameters[0]).IsNullChecked);
+        }
+
+        [Fact]
+        public void NullCheckedMethodValidationWithOptionalParameterSplitBySpace()
+        {
+            var source = @"
+class C
+{
+    void M(string name! =null) { }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+            var m = comp.GlobalNamespace.GetTypeMember("C").GetMember<SourceMethodSymbol>("M");
+            Debug.Assert(((SourceParameterSymbol)m.Parameters[0]).IsNullChecked);
+        }
+
+        [Fact]
+        public void NullCheckedConstructor()
+        {
+            var source = @"
+class C
+{
+    public C(string name!) { }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+            var m = comp.GlobalNamespace.GetTypeMember("C").GetMember<SourceMethodSymbol>(".ctor");
+            Debug.Assert(((SourceParameterSymbol)m.Parameters[0]).IsNullChecked);
+        }
+
+        [Fact]
+        public void NullCheckedOperator()
+        {
+            var source = @"
+class Box 
+{ 
+    public static int operator+ (Box b!, Box c)  
+    { 
+        return 2;
+    }
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+            var m = comp.GlobalNamespace.GetTypeMember("Box").GetMember<SourceUserDefinedOperatorSymbol>("op_Addition");
+            Debug.Assert(((SourceParameterSymbol)m.Parameters[0]).IsNullChecked);
+            Debug.Assert(!((SourceParameterSymbol)m.Parameters[1]).IsNullChecked);
+        }
+
+        [Fact]
+        public void NullCheckedLambdaParameter()
+        {
+            var source = @"
+using System;
+class C
+{
+    public void M()
+    {
+        Func<string, string> func1 = x! => x + ""1"";
+    }
+}";
+            var tree = Parse(source);
+            var comp = CreateCompilation(tree);
+            comp.VerifyDiagnostics();
+
+            CSharpSemanticModel model = (CSharpSemanticModel)comp.GetSemanticModel(tree);
+            var m = comp.GlobalNamespace.GetTypeMember("C").GetMember<SourceMethodSymbol>("M");
+            SimpleLambdaExpressionSyntax node = (SimpleLambdaExpressionSyntax)m.GetNonNullSyntaxNode().DescendantNodes().AsImmutable()[11];
```

Comment:
Consider changing to something like `m.GetNonNullSyntaxNode().DescendantNodes().OfType<SimpleLambdaExpressionSyntax>().Single()`

Reply:
Not a very important change, but maybe more resilient than having to change the number `11` to something else if the test ever needs to change.

Reply:
I thought we weren't allowed to use Linq?

Reply:
It's discouraged in implementation code but used somewhat frequently in tests.